Wrapper first steps

hendrikgit opened the pull request:
For this one I have a few questions.

I figured a library should not call `quit` so I raise an exception instead. I wanted to keep it simple and not look for the most appropriate existing nim exception for each netcdf function so I created my own `NetcdfError`. Is that ok?

I have a lot of
```nim
let retval = someNetcdfProc(...)
if retval != 0: raise ...
```
in my code that's why I now came up with the `handleError` template. What do you think of this approach?

I created a type for NcId because later there will be VarId and DimId as well. I thought it would be nice to not be able to confuse them in the higher level code. Even though they are all cint one does not want to mix them up I figure and put a varid where a ncid was supposed to be. What do you think?

Lastly about the name of `open` proc. It is ambiguous with `io.open` so when using the library one has to currently do `netcdf.open("thefile.nc")`. Is that enough reason to rename it to `ncOpen` or something?

hendrikgit committed:
update deprecated types

hendrikgit committed:
basic wrapper setup, procs for open and close

hendrikgit commented:
I think my current handleError template is too complicated. It could be changed to
```nim
template handleError(retval: cint) =
  if retval != 0:
    raise newException(NetcdfError, "Error " & $retval & ": " & $strerror(retval))
```
I'll wait for your feedback though.

hendrikgit commented:
Sorry had another thought about this. I got the template for error handling idea from the simple write example. But that could have been used there only because of a translation from C. I realized a `proc handleError(e: cint) = ...` would work fine here.

I read a bit on the topic about what to prefer and I came upon two answers, use the simpler construct first or "it depends" and if it really matters then measure it.

So I'm still wondering what to do here but I would pick proc now I suppose?

Vindaar commented:
> Lastly about the name of open proc. It is ambiguous with io.open so when using the library one has to currently do netcdf.open("thefile.nc"). Is that enough reason to rename it to ncOpen or something?

Yeah, for such a common name, keeping the prefix might be helpful. `ncOpen` sounds good to me.

> I figured a library should not call quit so I raise an exception instead. I wanted to keep it simple and not look for the most appropriate existing nim exception for each netcdf function so I created my own NetcdfError. Is that ok?

Perfect!


### About the error handling template

You have to keep the following in mind: 
The point of error handling on the one hand is to not continue running the program in an undefined state *and* to provide the user of the library with a) a way to handle the errors in their code and b) if unhandled / unexpected errors happen to get decent information where the error came from and what it was about.

So for the former, using exceptions in general is a perfectly suitable approach (I'm not going to go into the whole exception / error code / `Result` discussion, hehe).
For the latter I consider the "provide good information" the most important aspect. For that reason in `nimhdf5` in particular I don't have any kind of error handling template. The reason for that is not that I dislike the idea, but rather that I want to provide specific exception messages that include:
- the function that was called in the HDF5 library
- related information about the user's inputs if important, e.g. input data shapes 

Therefore I just do manual exception handling for every HDF5 function call. What I typically try to do is to have a "low level wrapper" of a HDF5 library call that handles the error and provides some good exception message. And in the "higher level" part of the code I use my wrapper function. 

*However*, in particular as a form of convenience to just get stuff done *I like* the idea of a generic error handling template. What it allows for is to avoid falling into the trap of just writing `discard netcdfCall`, because one is too lazy to write error handling on the spot (yes, I'm guilty of doing that sometimes...). 

While I agree with you that:
```nim
template handleError(retval: cint) =
  if retval != 0:
    raise newException(NetcdfError, "Error " & $retval & ": " & $strerror(retval))
```
is the simpler of the two approaches, the other variant:
```nim
template handleError(body: untyped) =
  let retval: cint = body
  if retval != 0:
    raise newException(NetcdfError, "Error " & $retval & ": " & $strerror(retval))
```
allows for a neat feature, if extended. Namely, consider this version:
```nim
template handleError(body: untyped) =
  let retval: cint = body
  if retval != 0:
    raise newException(NetcdfError, "Error " & $retval & ": " & $strerror(retval) & ". Call: " & astToStr(body)
```

Now it also includes the actual call that produced the error! The latter is really nice to combine the "convenient error handling for lazy people" with decent information about what call in the code actually caused the error.

So my tl;dr would be:
- error template is a good *default*, preferring the version taking a body to extend auto error message
- later probably a good idea to extend for more specific errors

Note: once one wants to replace individual error handling cases by specific ones, one can of course use your first error handling template and just give it an additional `message` argument that includes a custom error message containing useful information relevant in the calling scope.


Hope this clarified my thoughts a bit, haha.

Vindaar commented on src/netcdf.nim:
```diff
@@ -1,2 +1,20 @@
+
+proc `$`*(ncid: NcId): string = $ncid.int
+
+proc open*(path: string, mode = NC_NOWRITE): NCid =
```
I'm not familiar with the possible `mode` options `netCDF` provides. In HDF5 multiple "mode" like arguments may be combined. In that case I define an optional (more on that below) `set` of possible "mode" like flags.
That `set` contains elements of an `enum` that maps enum fields to the correct HDF5 constants. The constants there can be combined via bitwise OR operations, which one can conveniently do by OR'ing all elements of the set.

It is *optional*, because for more convenience I provide a default string like `"r"`, `"rw"` argument that is probably more expected for the average user coming from in case of HDF5 something like `h5py`. 

Not sure how this applies in `netCDF` though.

hendrikgit commented:
Thank you for your detailed feedback! I force pushed to the branch this PR is based on.
The template is the version with the AST for the body, I renamed open to ncOpen and added an enum for the open modes.

Vindaar commented:
Thanks again!
